crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_structure.rs

```diff
@@ -166,108 +155,129 @@ mod tests {
+        }
 
-        let width_of_row = full_width_of_row(13);
-        assert_eq!(width_of_row, 16);
+        // Verify the widths.
+        for (row, width) in expected_widths.iter().enumerate() {
+            let width_of_row = full_width_of_row(row);
+            assert_eq!(
+                width_of_row, *width,
+                "row: {} does not produce expected width",
+                row
+            );
+        }
     }
     #[test]
     fn we_can_produce_the_correct_matrix_size() {
-        let (width, height) = matrix_size(2, 0);
-        assert_eq!(width, 2);
-        assert_eq!(height, 2);
-
-        let (width, height) = matrix_size(3, 0);
-        assert_eq!(width, 2);
-        assert_eq!(height, 3);
-
-        let (width, height) = matrix_size(4, 0);
-        assert_eq!(width, 2);
-        assert_eq!(height, 3);
-
-        let (width, height) = matrix_size(16, 0);
-        assert_eq!(width, 4);
-        assert_eq!(height, 6);
+        // NOTE: returned tuple is (height, width).
+        assert_eq!(matrix_size(0, 0), (0, 0));
+        assert_eq!(matrix_size(1, 0), (1, 1));
+        assert_eq!(matrix_size(2, 0), (2, 2));
+        assert_eq!(matrix_size(3, 0), (3, 2));
+        assert_eq!(matrix_size(4, 0), (3, 2));
+        assert_eq!(matrix_size(5, 0), (4, 4));
+        assert_eq!(matrix_size(6, 0), (4, 4));
+        assert_eq!(matrix_size(7, 0), (4, 4));
+        assert_eq!(matrix_size(8, 0), (4, 4));
+        assert_eq!(matrix_size(9, 0), (5, 4));
+        assert_eq!(matrix_size(10, 0), (5, 4));
+        assert_eq!(matrix_size(11, 0), (5, 4));
+        assert_eq!(matrix_size(12, 0), (5, 4));
+        assert_eq!(matrix_size(13, 0), (6, 4));
+        assert_eq!(matrix_size(14, 0), (6, 4));
+        assert_eq!(matrix_size(15, 0), (6, 4));
+        assert_eq!(matrix_size(16, 0), (6, 4));
+        assert_eq!(matrix_size(17, 0), (7, 8));
 
-        let (width, height) = matrix_size(17, 0);
-        assert_eq!(width, 8);
-        assert_eq!(height, 7);
-
-        let (width, height) = matrix_size(64, 0);
-        assert_eq!(width, 8);
-        assert_eq!(height, 12);
-
-        let (width, height) = matrix_size(71, 0);
-        assert_eq!(width, 16);
-        assert_eq!(height, 13);
+        assert_eq!(matrix_size(64, 0), (12, 8));
+        assert_eq!(matrix_size(71, 0), (13, 16));
+        assert_eq!(matrix_size(81, 0), (14, 16));
+        assert_eq!(matrix_size(98, 0), (15, 16));
+        assert_eq!(matrix_size(115, 0), (16, 16));
     }
     #[test]
     fn we_can_produce_the_correct_matrix_size_with_offset() {
-        let (width, height) = matrix_size(2, 1);
-        assert_eq!(width, 2);
-        assert_eq!(height, 3);
-
-        let (width, height) = matrix_size(16, 1);
-        assert_eq!(width, 8);
-        assert_eq!(height, 7);
+        // NOTE: returned tuple is (height, width).
+        for i in 0..100 {
+            assert_eq!(matrix_size(0, i), (0, 0));
+        }
```

Comment:
See my point above. This is strange behavior, considering that `matrix_size(a, b)==matrix_size(c, d)` iff `a+b==c+d` in every case except this one.

Reply:
The code has been updated to be `matrix_size(a, b)==matrix_size(c, d)` iff `a+b==c+d` in every case.